### PR TITLE
extend xprotocol to multiple implementations

### PR DESCRIPTION
**Pull request by cch123**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/stream/xprotocol/connpool.go [hunk 1]
```diff
@@ -19,75 +19,63 @@ package xprotocol
...
-type connPool struct {
-	activeClients sync.Map //sub protocol -> activeClient
-	host          atomic.Value
-	mux           sync.Mutex
-	tlsHash       *types.HashValue
+// nolint
+type connpool struct {
+	host             atomic.Value
+	tlsHash          *types.HashValue
+	totalClientCount uint64 // total clients
```
Comment (nejisama) [inline pkg/stream/xprotocol/connpool.go:48]: totalClientCount is not counted in multiplex

**Review by nejisama — APPROVED** [review]

File: pkg/stream/xprotocol/connpool.go [hunk 2]
```diff
@@ -96,268 +84,30 @@ func (p *connPool) Host() types.Host {
...
-	if subProtocol != "" {
-		// check heartbeat enable, hack: judge trigger result of Heartbeater
-		proto := xprotocol.GetProtocol(subProtocol)
-		if heartbeater, ok := proto.(xprotocol.Heartbeater); ok && heartbeater.Trigger(0) != nil {
-			// create keepalive
-			rpcKeepAlive := NewKeepAlive(codecClient, subProtocol, time.Second, 6)
-			rpcKeepAlive.StartIdleTimeout()
-			ac.keepAlive = &keepAliveListener{
-				keepAlive: rpcKeepAlive,
+func getProtocol(ctx context.Context) types.ProtocolName {
```
[the remark this answers is not in the document]
Comment (cch123) [reply pkg/stream/xprotocol/connpool.go:100]: we may merge http1 && http2 conn pool with this one in the future, so the protocol should be a param

File: pkg/stream/xprotocol/connpool.go (hunk 1 above)
Comment (cch123) [reply pkg/stream/xprotocol/connpool.go:48]: it's fixed 1

**Review by taoyuanyuan — APPROVED** [review]
LGTM

**Pull request merged**